File: app/src/main/java/com/scichart/examples/fragments/examples2d/createRealtimeCharts/RealTimeMinMaxAnnotationFragment.java
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2021. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// RealTimeMinMaxAnnotationFragment.java is part of SCICHART®, High Performance Scientific Charts
// For full terms and conditions of the license, see http://www.scichart.com/scichart-eula/
//
// This source code is protected by international copyright law. Unauthorized
// reproduction, reverse-engineering, or distribution of all or any portion of
// this source code is strictly prohibited.
//
// This source code contains confidential and proprietary trade secrets of
// SciChart Ltd., and should at no time be copied, transferred, sold,
// distributed or made available without express written permission.
//******************************************************************************

package com.scichart.examples.fragments.examples2d.createRealtimeCharts;

import android.graphics.Color;
import android.os.Bundle;

import androidx.annotation.NonNull;

import com.scichart.charting.model.dataSeries.XyDataSeries;
import com.scichart.charting.visuals.SciChartSurface;
import com.scichart.charting.visuals.annotations.HorizontalAnchorPoint;
import com.scichart.charting.visuals.annotations.TextAnnotation;
import com.scichart.charting.visuals.annotations.VerticalAnchorPoint;
import com.scichart.charting.visuals.axes.AutoRange;
import com.scichart.charting.visuals.axes.NumericAxis;
import com.scichart.charting.visuals.renderableSeries.FastLineRenderableSeries;
import com.scichart.core.framework.UpdateSuspender;
import com.scichart.data.model.DoubleRange;
import com.scichart.data.model.ISciList;
import com.scichart.drawing.common.FontStyle;
import com.scichart.drawing.common.SolidPenStyle;
import com.scichart.examples.fragments.base.ExampleSingleChartBaseFragment;

import java.util.Random;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

public class RealTimeMinMaxAnnotationFragment extends ExampleSingleChartBaseFragment {

    private static final int FIFO_CAPACITY = 50;
    private static final long TIME_INTERVAL = 30;
    private static final double ONE_OVER_TIME_INTERVAL = 1.0 / TIME_INTERVAL;
    private static final double VISIBLE_RANGE_MAX = FIFO_CAPACITY * ONE_OVER_TIME_INTERVAL;
    private static final double GROW_BY = VISIBLE_RANGE_MAX * 0.1;

    private final Random random = new Random();

    private final XyDataSeries<Double, Double> ds1 = new XyDataSeries<>(Double.class, Double.class);
    private final DoubleRange xVisibleRange = new DoubleRange(GROW_BY, VISIBLE_RANGE_MAX + GROW_BY);
    private final DoubleRange yVisibleRange = new DoubleRange(0.0, 1.0);
    private final ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
    private ScheduledFuture<?> schedule;
    private TextAnnotation minAnnotation;
    private TextAnnotation maxAnnotation;
    private volatile boolean isRunning = true;
    private double t = 0.0;

    @Override
    protected void initExample(@NonNull SciChartSurface surface) {
        ds1.setFifoCapacity(FIFO_CAPACITY);

        UpdateSuspender.using(surface, () ->{
            NumericAxis xAxis = sciChartBuilder.newNumericAxis().withVisibleRange(xVisibleRange).withAutoRangeMode(AutoRange.Never).build();
            NumericAxis yAxis = sciChartBuilder.newNumericAxis().withVisibleRange(yVisibleRange).withAutoRangeMode(AutoRange.Never).build();

            FastLineRenderableSeries lineSeries = sciChartBuilder.newLineSeries().withDataSeries(ds1).withStrokeStyle(new SolidPenStyle(0xFF634e96, true, 2f, null)).build();

            surface.getXAxes().add(xAxis);
            surface.getYAxes().add(yAxis);
            surface.getRenderableSeries().add(lineSeries);

            minAnnotation = sciChartBuilder.newTextAnnotation()
                    .withText("Min: 0.0")
                    .withX1(5.0)
                    .withY1(5.0)
                    .withHorizontalAnchorPoint(HorizontalAnchorPoint.Center)
                    .withVerticalAnchorPoint(VerticalAnchorPoint.Bottom)
                    .withFontStyle(new FontStyle(40f, Color.RED))
                    .build();

            maxAnnotation = sciChartBuilder.newTextAnnotation()
                    .withText("Max: 0.0")
                    .withX1(0.0)
                    .withY1(0.0)
                    .withHorizontalAnchorPoint(HorizontalAnchorPoint.Center)
                    .withVerticalAnchorPoint(VerticalAnchorPoint.Top)
                    .withFontStyle(new FontStyle(40f, Color.GREEN))
                    .build();

            surface.getAnnotations().add(minAnnotation);
            surface.getAnnotations().add(maxAnnotation);
        });

        schedule = scheduledExecutorService.scheduleWithFixedDelay(insertRunnable, 0, TIME_INTERVAL, TimeUnit.MILLISECONDS);
    }

    private final Runnable insertRunnable = new Runnable() {
        @Override
        public void run() {
            if (!isRunning) return;


            double y1 = 3.0 * Math.sin(2 * Math.PI * 1.4 * t) + random.nextDouble() * 0.5;
            double y2 = 2.0 * Math.cos(2 * Math.PI * 0.8 * t) + random.nextDouble() * 0.5;
            double y3 = Math.sin(2 * Math.PI * 2.2 * t) + random.nextDouble() * 0.5;
            double y4 = Math.sin(2 * Math.PI * 2.2 * t) + random.nextDouble() * 0.5;

            ds1.append(t, y1 + y2 + y3 + y4);

            t += ONE_OVER_TIME_INTERVAL;
            if (t > VISIBLE_RANGE_MAX) {
                xVisibleRange.setMinMax(xVisibleRange.getMin() + ONE_OVER_TIME_INTERVAL, xVisibleRange.getMax() + ONE_OVER_TIME_INTERVAL);
            }

            // Update annotations with new min and max values
            double minY = ds1.getYRange().getMinAsDouble();
            double maxY = ds1.getYRange().getMaxAsDouble();
            int minIndex = ds1.getYValues().indexOf(minY);
            int maxIndex = ds1.getYValues().indexOf(maxY);
            double minX = ds1.getXValues().get(minIndex);
            double maxX = ds1.getXValues().get(maxIndex);
            double minAnnotationY = minY - (maxY - minY) * 0.02; // Move min annotation higher by 5% of the range
            double maxAnnotationY = maxY + (maxY - minY) * 0.02;

            minAnnotation.setText(String.format("Min: %.2f", minY));
            minAnnotation.setX1(minX);
            minAnnotation.setY1(minAnnotationY);

            maxAnnotation.setText(String.format("Max: %.2f", maxY));
            maxAnnotation.setX1(maxX);
            maxAnnotation.setY1(maxAnnotationY);


            // Calculate the adjusted range
            // range adjustment = data Range / (axis length - 2 * annotation height) * annotation height.
            double dr = (maxY - minY) / (binding.surface.getYAxes().get(0).getCurrentCoordinateCalculator().getViewportDimension() - 80) * 40;
            yVisibleRange.setMinMax(minY - dr, maxY + dr);

        }
    };

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        isRunning = false;

        outState.putDouble("time", t);
        outState.putParcelable("xValues1", ds1.getXValues());
        outState.putParcelable("yValues1", ds1.getYValues());
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        if (savedInstanceState != null) {
            t = savedInstanceState.getDouble("time");
            ds1.append((ISciList<Double>) savedInstanceState.getParcelable("xValues1"), (ISciList<Double>) savedInstanceState.getParcelable("yValues1"));
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        schedule.cancel(true);
    }

    private void resetChart() {
        UpdateSuspender.using(binding.surface, () -> {
            ds1.clear();

        });
    }
}

File: app/src/main/java/com/scichart/examples/fragments/examples2d/createRealtimeCharts/RealTimeMinMaxAnnotationFragment.kt
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2021. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// RealTimeMinMaxAnnotationFragment.kt is part of SCICHART®, High Performance Scientific Charts
// For full terms and conditions of the license, see http://www.scichart.com/scichart-eula/
//
// This source code is protected by international copyright law. Unauthorized
// reproduction, reverse-engineering, or distribution of all or any portion of
// this source code is strictly prohibited.
//
// This source code contains confidential and proprietary trade secrets of
// SciChart Ltd., and should at no time be copied, transferred, sold,
// distributed or made available without express written permission.
//******************************************************************************

package com.scichart.examples.fragments.examples2d.createRealtimeCharts.kt

import android.graphics.Color
import android.os.Bundle
import android.util.Log
import androidx.fragment.app.Fragment
import com.scichart.charting.visuals.SciChartSurface
import com.scichart.charting.visuals.annotations.HorizontalAnchorPoint
import com.scichart.charting.visuals.annotations.TextAnnotation
import com.scichart.charting.visuals.annotations.VerticalAnchorPoint
import com.scichart.charting.visuals.axes.AutoRange
import com.scichart.data.model.DoubleRange
import com.scichart.data.model.ISciList
import com.scichart.drawing.common.FontStyle
import com.scichart.examples.fragments.base.ExampleSingleChartBaseFragment

import com.scichart.examples.utils.scichartExtensions.SolidPenStyle
import com.scichart.examples.utils.scichartExtensions.XyDataSeries
import com.scichart.examples.utils.scichartExtensions.fastLineRenderableSeries
import com.scichart.examples.utils.scichartExtensions.numericAxis
import com.scichart.examples.utils.scichartExtensions.renderableSeries
import com.scichart.examples.utils.scichartExtensions.suspendUpdates
import com.scichart.examples.utils.scichartExtensions.xAxes
import com.scichart.examples.utils.scichartExtensions.yAxes
import java.util.Random
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import kotlin.math.cos
import kotlin.math.sin

class RealTimeMinMaxAnnotationFragment : ExampleSingleChartBaseFragment() {

    private val random = Random()

    private val ds1 = XyDataSeries<Double, Double>().apply { fifoCapacity =
        FIFO_CAPACITY
    }

    private val xVisibleRange = DoubleRange(GROW_BY, VISIBLE_RANGE_MAX + GROW_BY)
    private val yVisibleRange = DoubleRange(0.0, 1.0)

    private val scheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
    private lateinit var schedule: ScheduledFuture<*>
    private lateinit var minAnnotation: TextAnnotation
    private lateinit var maxAnnotation: TextAnnotation
    @Volatile
    private var isRunning = true

    override fun initExample(surface: SciChartSurface) {
        surface.suspendUpdates {
            xAxes { numericAxis {
                visibleRange = xVisibleRange
                autoRange = AutoRange.Never }
            }
            yAxes { numericAxis {
                //growBy = DoubleRange(0.1, 0.1)
                visibleRange = yVisibleRange
                autoRange = AutoRange.Never
            }
            }
            renderableSeries {
                fastLineRenderableSeries { dataSeries = ds1; strokeStyle = SolidPenStyle(0xFF634e96, 2f) }

            }
        }
        minAnnotation = sciChartBuilder.newTextAnnotation()
            .withText("Min: 0.0")
            .withX1(5.0)
            .withY1(5.0)
            .withHorizontalAnchorPoint(HorizontalAnchorPoint.Center)
            .withVerticalAnchorPoint(VerticalAnchorPoint.Bottom)
            .withFontStyle(FontStyle(40F, Color.RED))
            .build()

        maxAnnotation = sciChartBuilder.newTextAnnotation()
            .withText("Max: 0.0")
            .withX1(0.0)
            .withY1(0.0)
            .withHorizontalAnchorPoint(HorizontalAnchorPoint.Center)
            .withVerticalAnchorPoint(VerticalAnchorPoint.Top)
            .withFontStyle(FontStyle(40f, Color.GREEN))
            .build()

        surface.annotations.add(minAnnotation)
        surface.annotations.add(maxAnnotation)
        schedule = scheduledExecutorService.scheduleWithFixedDelay(insertRunnable, 0, TIME_INTERVAL, TimeUnit.MILLISECONDS)
    }

    var t = 0.0
    private val insertRunnable = Runnable {
        if (!isRunning) return@Runnable

        binding.surface.suspendUpdates {
            val y1 = 3.0 * sin(2 * Math.PI * 1.4 * t) + random.nextDouble() * 0.5
            val y2 = 2.0 * cos(2 * Math.PI * 0.8 * t) + random.nextDouble() * 0.5
            val y3 = sin(2 * Math.PI * 2.2 * t) + random.nextDouble() * 0.5
            val y4 = sin(2 * Math.PI * 2.2 * t) + random.nextDouble() * 0.5
            ds1.append(t, y1+y2+y3+y4)


            t += ONE_OVER_TIME_INTERVAL
            if (t > VISIBLE_RANGE_MAX) {
                xVisibleRange.setMinMax(xVisibleRange.min + ONE_OVER_TIME_INTERVAL, xVisibleRange.max + ONE_OVER_TIME_INTERVAL)
            }
            // Update annotations with new min and max values
            val minY = ds1.yRange.minAsDouble
            val maxY = ds1.yRange.maxAsDouble
            val minIndex = ds1.yValues.indexOf(minY)
            val maxIndex = ds1.yValues.indexOf(maxY)
            val minX = ds1.xValues[minIndex]
            val maxX = ds1.xValues[maxIndex]
            val minAnnotationY = minY - (maxY - minY) * 0.02// Move min annotation higher by 5% of the range
            val maxAnnotationY = maxY + (maxY - minY) * 0.02
            // Log values to debug
            Log.d("SciChart", "Min: x=$minX, y=$minY, Max: x=$maxX, y=$maxY")

            minAnnotation.text = String.format("Min: %.2f", minY)
            minAnnotation.x1 = minX
            minAnnotation.y1 = minAnnotationY

            maxAnnotation.text = String.format("Max: %.2f", maxY)
            maxAnnotation.x1 = maxX
            maxAnnotation.y1 = maxAnnotationY

            // Calculate the adjusted range
            // range adjustment = data Range / (axis length - 2 * annotation height) * annotation height.
            val dr = (maxY - minY) / (binding.surface.yAxes[0].currentCoordinateCalculator.viewportDimension - 80) * 40
            yVisibleRange.setMinMax(minY - dr, maxY + dr)

        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        isRunning = false

        outState.run {
            putDouble("time", t)
            putParcelable("xValues1", ds1.xValues)
            putParcelable("yValues1", ds1.yValues)

        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        savedInstanceState?.run {
            t = getDouble("time")
            val xValues1: ISciList<Double> = getParcelable("xValues1")!!
            val yValues1: ISciList<Double> = getParcelable("yValues1")!!
            ds1.append(xValues1, yValues1)

        }
    }

    override fun onDestroyView() {
        super.onDestroyView()

        schedule.cancel(true)
    }
    private fun resetChart() {
        binding.surface.suspendUpdates {
            ds1.clear()
        }
    }

    companion object {
        private const val FIFO_CAPACITY = 50
        private const val TIME_INTERVAL: Long = 30
        private const val ONE_OVER_TIME_INTERVAL = 1.0 / TIME_INTERVAL
        private const val VISIBLE_RANGE_MAX = FIFO_CAPACITY * ONE_OVER_TIME_INTERVAL
        private const val GROW_BY = VISIBLE_RANGE_MAX * 0.1
    }
}
